llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -7687,6 +7687,30 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   if (Instruction *Res = foldReductionIdiom(I, Builder, DL))
     return Res;
 
+  {
+    Value *A;
+    const APInt *C1, *C2;
+    ICmpInst::Predicate PredEq = ICmpInst::ICMP_EQ;
+    if (I.getPredicate() == PredEq) {
+      // c2 must be positive
+      // sext(a) & c1 == c2 --> a & c3 == trunc(c2)
+      if (match(Op0, m_And(m_SExtLike(m_Value(A)), m_APInt(C1))) && match(Op1, m_APInt(C2))) {
+        if (C2->isNonNegative()) {
```

Comment:
`c2` must be non-negative at the bitwidth of `a`. For example, if `a` is an 8-bit integer and `c2` is an 32-bit integer, the leading 25 bits of `c2` should be zeros.
Here you are actually checking if `c2` is non-negative at the bitwidth of `c2`.

Reply:
Or just `C2.ult(APInt::getOneBitSet(C2.getBitWidth(), InputBitWidth)`